Src/Newtonsoft.Json.Bson/BsonDataReader.Async.cs

```diff
@@ -0,0 +1,1018 @@
+        }
+
+        private Task<byte[]> ReadBytesAsync(int count, CancellationToken cancellationToken)
+        {
+            MovePosition(count);
+            return _asyncReader.ReadBytesAsync(count, cancellationToken);
+        }
+
+        private async Task<JsonToken> GetContentTokenAsync(CancellationToken cancellationToken)
+        {
+            while (await ReadAsync(cancellationToken).ConfigureAwait(false))
+            {
+                JsonToken t = TokenType;
+                if (t != JsonToken.Comment)
+                {
+                    return t;
+                }
+            }
+
+            SetToken(JsonToken.None);
+            return JsonToken.None;
+        }
+
+        /// <summary>
+        /// Asynchronously reads the next JSON token from the source as a <see cref="T:System.Nullable`1" /> of <see cref="T:System.Boolean" />.
+        /// </summary>
+        /// <param name="cancellationToken">The token to monitor for cancellation requests. The default value is <see cref="P:System.Threading.CancellationToken.None" />.</param>
+        /// <returns>A <see cref="T:System.Threading.Tasks.Task`1" /> that represents the asynchronous read. The <see cref="P:System.Threading.Tasks.Task`1.Result" />
+        /// property returns the <see cref="T:System.Nullable`1" /> of <see cref="T:System.Boolean" />. This result will be <c>null</c> at the end of an array.</returns>
+        /// <remarks>Derived classes must override this method to get asynchronous behaviour. Otherwise it will
+        /// execute synchronously, returning an already-completed task. Asynchronous behaviour is also not available when the
+        /// constructor was passed an instance of type derived from <see cref="BinaryReader"/>.</remarks>
+        public override Task<bool?> ReadAsBooleanAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            return _asyncReader == null ? base.ReadAsBooleanAsync(cancellationToken) : DoReadAsBooleanAsync(cancellationToken);
+        }
+
+        private async Task<bool?> DoReadAsBooleanAsync(CancellationToken cancellationToken)
+        {
+            JsonToken t = await GetContentTokenAsync(cancellationToken).ConfigureAwait(false);
+
+            switch (t)
+            {
+                case JsonToken.None:
+                case JsonToken.Null:
+                case JsonToken.EndArray:
+                    return null;
+                case JsonToken.Integer:
+                case JsonToken.Float:
+                    bool b;
+#if HAVE_BIG_INTEGER
+                    if (Value is BigInteger)
+                    {
+                        b = (BigInteger)Value != 0;
+                    }
+                    else
+#endif
+                    {
+                        b = Convert.ToBoolean(Value, CultureInfo.InvariantCulture);
+                    }
+
+                    SetToken(JsonToken.Boolean, b, false);
```

Comment:
One thing that currently needs to acccess internals, though maybe a protected `ReadCurrentTokenAs` set of methods could efficiently serve here, since it's `GetContentTokenAsync` that is doing the actual asynchronous reading.